internal/gogrep/match.go

```diff
@@ -0,0 +1,731 @@
+package gogrep
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"strconv"
+
+	"github.com/go-toolsmith/astequal"
+)
+
+type matcher struct {
+	prog *program
+
+	insts []instruction
+	pc    int
+
+	// node values recorded by name, excluding "_" (used only by the
+	// actual matching phase)
+	capture []CapturedNode
+}
+
+func newMatcher(prog *program) *matcher {
+	return &matcher{
+		prog:    prog,
+		insts:   prog.insts,
+		capture: make([]CapturedNode, 0, 8),
+	}
+}
+
+func (m *matcher) nextInst() instruction {
+	inst := m.insts[m.pc]
+	m.pc++
+	return inst
+}
+
+func (m *matcher) stringValue(inst instruction) string {
+	return m.prog.strings[inst.valueIndex]
+}
+
+func (m *matcher) ifaceValue(inst instruction) interface{} {
+	return m.prog.ifaces[inst.valueIndex]
+}
+
+func (m *matcher) MatchNode(n ast.Node, accept func(MatchData)) {
+	m.pc = 0
+	inst := m.nextInst()
+	switch {
+	case inst.op == opMultiStmt:
```

Comment:
Should `inst.op` be in `switch` expr?

Reply:
Good point.

Reply:
When gocritic will start to suggest this kind of rewrite? :sweat_smile:

Reply:
Fixed.